text/0000-century-of-the-self-type.md

```diff
@@ -0,0 +1,169 @@
+
+For example, this self type is valid:
+
+```rust
+impl MyType {
+     fn by_ref_to_rc(self: &Rc<Self>) { ... }
+}
+```
+
+## Object safety
+
+In order for these receivers to be object safe, some additional traits need to
+be implemented. Given a reference type `Ptr<dyn Trait>`, the compiler must be
+able to prove that `T: Unsize<dyn Trait>` implies `Ptr<T>:
+CoerceUnsized<Ptr<dyn Trait>>`. If the compiler can prove this, methods with
+these receivers are object safe (how they object safe conversion is implemented
+is discussed later in the detailed design).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Handling trait objects
+
+To support object-safety, we must have a way of obtaining a vtable from the
+reference type, and then passing the correct receiver type to the function in 
+the vtable.
+
+### Step 1: Obtaining the vtable
+
+First, we call `Deref::deref` on the receiver type to obtain an `&Trait`
+object. This will return an `&dyn Trait`. Since an `&dyn Trait` is defined as 2
+words, one a pointer to the object data and the other a pointer to the vtable,
+we can obtain the vtable by dereferencing the second pointer of this object.
+
+For example, consider this type:
+
+```rust
+struct Foo<T: ?Sized> {
+    inner: T,
+}
+
+impl<T: ?Sized> Deref for Foo<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.inner
+    }
+}
+
+trait Bar {
+    fn baz(self: Foo<Self>);
+}
+```
+
+Here, we start with an `&Foo<Trait>`, which is represented as a wide pointer.
+When we call `Deref::deref`, we receive a wide pointer, with the data pointer
+pointing to `self.inner`, and the vtable pointing to that same vtable.
+
+### Step 2: Obtaining the correct receiver type
+
+Having obtained the vtable, we now need to obtain a value of the correct type
+to pass to the function. For example, given a trait like this:
+
+```rust
+trait Foo {
+    fn bar(self: Rc<Self>);
+}
+```
+
+The function in the vtable expects a value of `Rc<Self>`, where `Self` is the
+concrete type that was cast into the vtable. So if we have an `Rc<Foo>`, we
+need to temporarily cast that *back* into an `Rc<i32>` or whatever the concrete
+type is.
+
+This is why the `CoerceUnsized` bound is necessary for object-safe receiver
+types. Using the type ID stored in the vtable, we can downcast the `Self` type
+to the correct concrete type to pass to the function pointer for the method
+we're calling, effectively reversing the unsizing coercion.
+
+## Stabilization plans
+
+As soon as possible, we intend to stabilize *using* a method receiver defined
+this way, so long as it is object safe. That stabilization will immediately
+extend support to `Rc<T>` and `Arc<T>`.
+
+However, we also feel that `CoerceUnsized` is not ready to stabilize without
+further consideration of the trade offs. For that reason, defining your own
+arbitrary method receivers may not be stabilized as quickly.
```

Comment:
I think this division makes sense. It shouldn't even require an extra feature gate, since `CoerceUnsized` is already unstable, correct? In other words, the delayed stabilization described here doesn't require delaying stabilization of any part of this RFC-- it just specifies that `CoerceUnsized`, an existing unstable feature, will remain unstable.